docs/design/resource-qos.md

```diff
@@ -0,0 +1,224 @@
+- The base OOM score is between 0 and 1000, so if process A’s OOM_SCORE_ADJ - process B’s OOM_SCORE_ADJ is over a 1000, then process A will always be OOM killed before B.
+- The final OOM score of a process is also between 0 and 1000
+
+*Best-effort*
+	- Set OOM_SCORE_ADJ: 1000
+	- So processes in best-effort containers will have an OOM_SCORE of 1000
+
+*Guaranteed*
+	- Set OOM_SCORE_ADJ: -999
+	- So processes in guaranteed containers will have an OOM_SCORE of 0 or 1
+
+*Burstable*
+	- If total memory request > 99.8% of available memory, OOM_SCORE_ADJ: 2
+	- Otherwise, set OOM_SCORE_ADJ to 1000 - 10 * (% of memory requested)
+	- This ensures that the OOM_SCORE of burstable containers is > 1
+	- So burstable containers will be killed if they conflict with guaranteed containers
+	- If a burstable container uses less memory than requested, its OOM_SCORE < 1000
+	- So best-effort containers will be killed if they conflict with burstable containers using less than requested memory
+	- If a process in a burstable container uses more memory than the container requested, its OOM_SCORE will be 1000, if not its OOM_SCORE will be < 1000
+	- Assuming that a container typically has a single big process, if a burstable container that uses more memory than requested conflicts with a burstable container using less memory than requested, the former will be killed
+	- If burstable containers with multiple processes conflict, then the formula for OOM scores is a heuristic, it will not ensure "Request and Limit" guarantees. This is one reason why control loops will be added in subsequent iterations.
+
+*Pod infrastructure container*
+  - OOM_SCORE_ADJ: -999
+*Kubelet, Docker, Kube-Proxy*
+  - OOM_SCORE_ADJ: -999 (won’t be OOM killed)
+  - Hack, because these critical tasks might die if they conflict with guaranteed containers. in the future, we should place all user-pods into a separate cgroup, and set a limit on the memory they can consume.
+
+## Known issues and possible improvements
+
+The above implementation provides for basic oversubscription with protection, but there are a few known limitations.
+
+#### Supporting other runtimes
+
+- *RKT*: The current design focuses on Docker. Support for other runtimes needs to be addressed.
+- *Systemd*: Systemd platforms need to be handled in a different way. Kubelet needs to be systemd aware and use systemd APIs for QoS management.
+
+#### Support for Swap
+
+- The current QoS policy assumes that swap is disabled. If swap is enabled, then resource guarantees (for pods that specify resource requirements) will not hold. For example, suppose 2 guaranteed pods have reached their memory limit. They can continue allocating memory by utilizing disk space. Eventually, if there isn’t enough swap space, processes in the pods might get killed. The node must take into account swap space explicitly for providing deterministic isolation behavior.
+
```

Comment:
Makes sense. Added a section on this. We can collect any future changes to the node for QoS in this doc.